File: day02/day02.py
# Day 02: Cube Conundrum

import string
import sys

# ==============================================================================

# If no specific input given, default to "day<X>.in"
if len(sys.argv) > 1:
  fname = sys.argv[1]
else:
  fname = sys.argv[0].replace(".py", ".in")

# Parse input
games = []
with open(fname) as f:
  for line in f:
    num = int(line.strip().split(":")[0].replace("Game ", ""))
    sets = []
    for _set in line.strip().split(":")[1].split(";"):
      draws = []
      draw = _set.split(",")
      for d in draw:
        amount = int(d.strip().split(" ")[0])
        color = d.strip().split(" ")[1]
        draws.append((amount, color))
      sets.append(draws)
    games.append((num, sets))     

# ------------------------------------------------------------------------------
# Part 1

max_cubes = {"red": 12, "green": 13, "blue": 14}

def is_valid(game):
  num, sets = game
  for _set in sets:
    for amount, color in _set:
      if amount > max_cubes[color]:
        return False
  return True

valids = []
for game in games:
  if is_valid(game):
    valids.append(game[0])

print("Part 1:", sum(valids))

# ------------------------------------------------------------------------------
# Part 2

tot_power = 0
for game in games:
  num, sets = game
  min_cubes = {"red": 0, "green": 0, "blue": 0}
  for _set in sets:
    for amount, color in _set:
      min_cubes[color] = max(amount, min_cubes[color])
  power = min_cubes["red"]*min_cubes["green"]*min_cubes["blue"]
  tot_power += power

print("Part 2:", tot_power)

File: day02/day02.jl
# Day 02: Cube Conundrum

# ==============================================================================

# Get input filename as CLI arg. If not given, default to "day<X>.in"
if length(ARGS) > 0
  fname = ARGS[1]
else
  fname = replace(PROGRAM_FILE, ".jl" => ".in")
end

# Parse input
games = []
lines = readlines(fname)
for line in lines
  part1, part2 = split(line, ":")
  game_num = parse(Int, part1[6:end])
  sets = []
  for _set in split(part2, ";")
    set = []
    for draw in split(_set, ",")
      amount, color = split(strip(draw), " ")
      push!(set, (parse(Int, amount), color))
    end
    push!(sets, set)
  end
  push!(games, (game_num, sets))
end

# ------------------------------------------------------------------------------
# Part 1

max_cubes = Dict("red"=>12, "green"=>13, "blue"=>14)

function is_valid(game)
  game_num, sets = game
  for set in sets
    for draw in set
      amount, color = draw
      if amount > max_cubes[color]
        return false
      end
    end
  end
  return true
end

valids = []
for game in games
  if is_valid(game)
    push!(valids, game[1])
  end
end

println("Part 1: ", sum(valids))

# ------------------------------------------------------------------------------
# Part 2

tot_power = 0
for game in games
  num, sets = game
  min_cubes = Dict("red"=>0, "green"=>0, "blue"=>0)
  for set in sets
    for draw in set
      amount, color = draw
      min_cubes[color] = max(amount, min_cubes[color])
    end
  end
  power = min_cubes["red"]*min_cubes["green"]*min_cubes["blue"]
  global tot_power += power
end

println("Part 2: ", tot_power)